File: aws/app/src/cognito-pre-auth.py
import boto3

def lambda_handler(event, context):
    client = boto3.client('cognito-idp')
    
    print(event)
    
    if event['request']['validationData'] == None:
        return event
    
    user_pool_id = event['userPoolId']
    username = event['userName']
    

    
    tenant_attribute_name = 'custom:active_tenant'
    tenant_attribute_value = event['request']['validationData']['tenant_id']
    
    role_attribute_name = 'custom:role'
    role_attribute_value = event['request']['validationData']['role']
    
    tenant_user_id_attribute_name = 'custom:tenant_user_id'
    tenant_user_id_attribute_value = event['request']['validationData'].get('tenant_user_id', '')
    
    if tenant_attribute_value == '' or role_attribute_value == '':
        return event
    
    try:
        # Update the user attribute
        user_attributes = [
            {
                'Name': tenant_attribute_name,
                'Value': tenant_attribute_value
            },
            {
                'Name': role_attribute_name,
                'Value': role_attribute_value
            }
        ]
        
        if tenant_user_id_attribute_value:
            user_attributes.append({
                'Name': tenant_user_id_attribute_name,
                'Value': tenant_user_id_attribute_value
            })
        
        response = client.admin_update_user_attributes(
            UserPoolId=user_pool_id,
            Username=username,
            UserAttributes=user_attributes
        )
        print(f"Attributes updated successfully for user {username}.")
    except client.exceptions.ClientError as error:
        print(f"Error updating user attribute: {error}")
        raise error
    
    # Continue the authentication process
    return event

File: aws/app/src/cognito-pre-auth.mjs
import { CognitoIdentityProviderClient, AdminUpdateUserAttributesCommand } from "@aws-sdk/client-cognito-identity-provider";

const client = new CognitoIdentityProviderClient({});

export const handler = async (event) => {
    console.log(JSON.stringify(event, null, 2));

    const validationData = event.request.validationData;

    // check if validation data exist
    if (!validationData) {
        return event;
    }

    const userPoolId = event.userPoolId;
    const username = event.userName;

    const tenantAttributeValue = validationData.tenant_id;
    const roleAttributeValue = validationData.role;
    const tenantUserIdAttributeValue = validationData.tenant_user_id || '';

    // check teh field
    if (!tenantAttributeValue || !roleAttributeValue) {
        return event;
    }

    try {
        const userAttributes = [
            {
                Name: 'custom:active_tenant',
                Value: tenantAttributeValue
            },
            {
                Name: 'custom:role',
                Value: roleAttributeValue
            }
        ];

        // add tenant_user_id, if it exist
        if (tenantUserIdAttributeValue) {
            userAttributes.push({
                Name: 'custom:tenant_user_id',
                Value: tenantUserIdAttributeValue
            });
        }

        const command = new AdminUpdateUserAttributesCommand({
            UserPoolId: userPoolId,
            Username: username,
            UserAttributes: userAttributes
        });

        await client.send(command);
        console.log(`Attributes updated successfully for user ${username}.`);

    } catch (error) {
        console.error(`Error updating user attribute: ${error.message}`);
        throw error;
    }

    // continue autentification
    return event;
};